components/scream/src/physics/p3/tests/p3_unit_tests.cpp

```diff
@@ -80,6 +81,474 @@ struct UnitWrap::UnitTest<D>::TestP3Func
+      rwdc_device(0).qrevp = qrevp[0];
+      rwdc_device(0).qrcol = qrcol[0];
+      rwdc_device(0).qrheti = qrheti[0];
+    });
+
+
+    // Sync back to host
+    Kokkos::deep_copy(rwdc_host, rwdc_device);
+    const auto ratio = rwdc[0].qr/(rwdc[0].qrevp * rwdc[0].dt);
+
+    //Here we check cases where source > sinks and sinks > 1e-20
+    REQUIRE(rwdc_host(0).qcaut == 0.0);
+    REQUIRE(rwdc_host(0).qcacc == 0.0);
+    REQUIRE(rwdc_host(0).qimlt == 0.0);
+    REQUIRE(rwdc_host(0).qcshd == 0.0);
+
+    //Check the value of qrevp
+    REQUIRE(abs(rwdc_host(0).qrevp- rwdc[0].qrevp*ratio)<= C::Tol);
+
+    //Now test that conservation has actually been enforced
+    REQUIRE( rwdc_host(0).qrevp * rwdc_host(0).dt  <= rwdc_host(0).qr);
+
+  }
+
+ 
+  static void ice_water_conservation_tests_device(){
+    using KTH = KokkosTypes<HostDevice>;
+
+    IceWaterConservationData iwdc[1] = {{sp(1e-5), 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, sp(1.1), sp(1e-4), 0.0}};
+
+    // Sync to device
+    KTH::view_1d<IceWaterConservationData> iwdc_host("iwdc_host", 1);
+    view_1d<IceWaterConservationData> iwdc_device("iwdc_host", 1);
+
+    // This copy only copies the input variables.
+    std::copy(&iwdc[0], &iwdc[0] + 1, iwdc_host.data());
+    Kokkos::deep_copy(iwdc_device, iwdc_host);
+
+    // Run the lookup from a kernel and copy results back to host
+    Kokkos::parallel_for(RangePolicy(0, 1), KOKKOS_LAMBDA(const Int& i) {
+      Spack qitot(iwdc_device(0).qitot);
+      Spack qidep(iwdc_device(0).qidep);
+      Spack qinuc(iwdc_device(0).qinuc);
+      Spack qrcol(iwdc_device(0).qrcol);
+      Spack qccol(iwdc_device(0).qccol);
+      Spack qrheti(iwdc_device(0).qrheti);
+      Spack qcheti(iwdc_device(0).qcheti);
+      Spack qiberg(iwdc_device(0).qiberg);
+      Spack qisub(iwdc_device(0).qisub);
+      Spack qimlt(iwdc_device(0).qimlt);
+
+      Functions::ice_water_conservation(qitot, qidep, qinuc, qrcol, qccol, qrheti, qcheti, qiberg, iwdc_device(0).dt, qisub, qimlt);
+
+      iwdc_device(0).qitot = qitot[0];
+      iwdc_device(0).qidep = qidep[0];
+      iwdc_device(0).qinuc = qinuc[0];
+      iwdc_device(0).qrcol = qrcol[0];
+      iwdc_device(0).qccol = qccol[0];
+      iwdc_device(0).qrheti = qrheti[0];
+      iwdc_device(0).qcheti = qcheti[0];
+      iwdc_device(0).qiberg = qiberg[0];
+      iwdc_device(0).qisub = qisub[0];
+      iwdc_device(0).qimlt = qimlt[0];
+    });
+
+  }
+
+  static void run()
+  {
+
+    cloud_water_conservation_tests_device();
+
+    rain_water_conservation_tests_device();
+
+    ice_water_conservation_tests_device();
+
+  }
+
+  static void cloud_water_conservation_unit_bfb_tests(){
+
+    using KTH = KokkosTypes<HostDevice>;
+
+    static constexpr Int max_pack_size = 16;
+    REQUIRE(Spack::n <= max_pack_size);
+
+    CloudWaterConservationData cwdc[max_pack_size] = {
+      {9.9999999999999995e-7, 0.0, 1800.0, 1.5832574016248739e-12, 1.0630996907148179e-12, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
```

Comment:
I've found it's helpful to label what each column is. That's the pattern we've established in the other tests.